clouddq/classes/bigquery_client.py

```diff
@@ -0,0 +1,177 @@
+        elif gcp_service_account_key_path:
+            source_credentials = service_account.Credentials.from_service_account_file(
+                filename=gcp_service_account_key_path, scopes=TARGET_SCOPES
+            )
+        # Otherwise, use Application Default Credentials
+        else:
+            source_credentials, _ = google.auth.default(
+                scopes=TARGET_SCOPES, quota_project_id=project_id
+            )
+        # Attempt to refresh token if not currently valid
+        if not source_credentials.valid:
+            auth_req = google.auth.transport.requests.Request()
+            source_credentials.refresh(auth_req)
+        # Attempt service account impersonation if requested
+        if gcp_impersonation_credentials:
+            target_credentials = impersonated_credentials.Credentials(
+                source_credentials=source_credentials,
+                target_principal=gcp_impersonation_credentials,
+                target_scopes=TARGET_SCOPES,
+                lifetime=500,
+            )
+            self.__credentials = target_credentials
+        else:
+            # Otherwise use source_credentials
+            self.__credentials = source_credentials
+        # Try to get service account credentials email
+        if self.__credentials.__dict__.get("_service_account_email"):
+            self.__user_email = self.__credentials.service_account_email
+        elif self.__credentials.__dict__.get("_target_principal"):
+            self.__user_email = self.__credentials.service_account_email
+        else:
+            # Otherwise try to get ADC credentials email
+            request = google.auth.transport.requests.Request()
+            token = self.__credentials.id_token
+            id_info = id_token.verify_oauth2_token(token, request)
+            self.__user_email = id_info["email"]
+        if self.__user_email:
+            logger.info(
+                f"Successfully created BigQuery Client with user: "
+                f"{self.__user_email}"
+            )
+        else:
+            logger.warning(
+                "Encountered error while retrieving user from ADC " "credentials.",
+            )
+        # Get project ID
+        if project_id:
+            self.__project_id = project_id
+        elif self.__credentials.__dict__.get("_project_id"):
+            self.__project_id = self.__credentials.project_id
+        elif self.__credentials.__dict__.get("_quota_project_id"):
+            self.__project_id = self.__credentials._quota_project_id
+        else:
+            logger.warn(
+                "Could not retrieve project_id from ADC credentials.", exc_info=True
+            )
+
+    def get_connection(self, new: bool = False) -> bigquery.client.Client:
+        """Creates return new Singleton database connection"""
+        if self.__client is None or new:
+            job_config = bigquery.QueryJobConfig(dry_run=True, use_query_cache=False)
```

Comment:
is dry run required? Also, can we make use_query_cache to True, so that it uses results from cache for same queries and it will be flushed once the table is modified.

Reply:
Hmm this was originally meant for the dry-run client. Maybe we can rename this to "get_dry_run_connection" and have a separate method called "get_connection" for non-dry-run queries? Let me make that update.

Reply:
I ended up removing the dry-run configs from `get_connection` and instead moving the dry-run job config to the method `check_query_dry_run` directly